### PR TITLE
Closed source infrastructure

### PR DESCRIPTION
**Pull request by maayanKash**
level testing.
#10852

### CODE REVIEW

**Comment by maayanKash** [comment]
@timothytrippel 
@sriyerg

File: meson_options.txt [hunk 1]
```diff
@@ -25,3 +25,16 @@ option(
...
+  type: 'boolean',
+  value: false,
+)
+
+option(
+  'partnerdir',
+  type: 'string',
+  value: '',
+)
+
```
Comment (timothytrippel) [inline meson_options.txt:40]: nit: remove new line

File: sw/device/lib/testing/test_framework/ottf.c [hunk 2]
```diff
@@ -54,11 +54,21 @@ static void report_test_status(bool result) {
...
 // A wrapper function is required to enable `test_main()` and test teardown
 // logic to be invoked as a FreeRTOS task. This wrapper can be used by tests
 // that are run on bare-metal.
 static void test_wrapper(void *task_parameters) {
-  bool result = test_main();
+  // Invoke weak symbol that can be overriden in closed source code.
+  bool result = closed_source_pre_test_hook();
+  
+  // Invoke weak symbol that can be overriden in closed source code. 
+  result = result && test_main() && closed_source_post_test_hook();
```
Comment (timothytrippel) [inline sw/device/lib/testing/test_framework/ottf.c:71]: Nit: the following is a little more readable, while still taking advantage of C's short-circuit expression evaluation
```suggestion
  result = result && test_main();
  
  // Invoke weak symbol that can be overriden in closed source code. 
  result = result && closed_source_post_test_hook();
```

File: sw/device/tests/closed_source/meson.build [hunk 3]
```diff
@@ -0,0 +1,37 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+# All build targets below in `sw_tests` or `sw_signed_tests` will have names
+# starting with `sw/device/tests/<test_name>`. They will not contain the
+# subdirectory name (e.g., `sim_dv/`), because the build targets are really
+# declared in the for loops in this build file.
+
```
Comment (timothytrippel) [inline sw/device/tests/closed_source/meson.build:9]: You can remove this. This comment is not applicable in this file (only the file from the directory above, i.e., `sw/device/tests/meson.build`). Moreover, I just realized this comment is outdated in the  parent directory (where it was copied from), so thanks for pointing that out :)

File: sw/device/tests/meson.build [hunk 4]
```diff
@@ -816,6 +816,13 @@ subdir('autogen')
 ###############################################################################
 subdir('sim_dv')
 
+if get_option('partner') == true
+###############################################################################
+# Closed Source Test Libraries
+###############################################################################
+subdir('closed_source')
+endif
+
```
Comment (timothytrippel) [inline sw/device/tests/meson.build:825]: nit: maybe the following is a bit more readable?
```suggestion
###############################################################################
# Closed Source Test Libraries
###############################################################################
if get_option('partner')
    subdir('closed_source')
endif
```

File: sw/device/tests/meson.build [hunk 5]
```diff
@@ -877,16 +884,33 @@ foreach sw_test_name, sw_test_info : sw_tests
...
-      sw_test_name + '_' + device_name,
-      name_suffix: 'elf',
-      dependencies: [
-        shared_test_deps,
-        sw_test_info['library'],
-      ],
-    )
-
+       
+    if get_option('partner') == true and \
```
Comment (timothytrippel) [inline sw/device/tests/meson.build:888]:
```suggestion
    if get_option('partner') and \
```

File: sw/device/tests/closed_source/meson.build (hunk 3 above)
Comment (timothytrippel) [reply sw/device/tests/closed_source/meson.build:9]: Moreover, maybe you can replace the comment you remove above with one stating that the test targets defined in this build file are only designed to build in the closed-source environement, i.e., when `partner` and `partnerdir` options are set?

File: meson_options.txt [hunk 6]
```diff
@@ -25,3 +25,16 @@ option(
   type: 'boolean',
   value: false,
 )
+
+option(
+  'partner',
```
Comment (timothytrippel) [inline meson_options.txt:30]: maybe `build_closed_source` is a more descriptive name?

File: meson_options.txt [hunk 7]
```diff
@@ -25,3 +25,16 @@ option(
...
+  type: 'boolean',
+  value: false,
+)
+
+option(
+  'partnerdir',
```
Comment (timothytrippel) [inline meson_options.txt:36]: same here, perhaps `closed_source_dir`? wdyt?

**Review by timothytrippel — COMMENTED** [review]
Thanks for submitting this @maayanKash ! I just added a few minor cosmetic suggestions, otherwise overall LGTM.

We will probably want to add this same support to the bazel build as well; maybe @drewmacrae has some suggestions on how to accomplish this? While adding bazel support probably makes sense to do in a separate PR, @maayanKash would you mind filing an issue related to this, so we can track it?

**Comment by timothytrippel** [comment]
> level testing. #10852

Two other things:

1) Could you please clean up your commit history on this PR, by doing an interactive rebase (`git rebase -i HEAD~7`) and squashing all the commits in to a single commit with a clean commit message? See [here](https://docs.opentitan.org/doc/ug/github_notes/) for how you should structure your commit message, and [here](https://github.com/lowRISC/opentitan/pull/11028/commits/a43521cc36bec57ba08c5a0ca1b68edf0f5037ff) for an actual example.

2) Could you also provide a good PR description beyond just a title? Once you clean up the commit message (described above), you can use this for the PR description, since this PR should only contain a single commit. 

Let me know if you have any questions!

**Comment by timothytrippel** [comment]
> > level testing. #10852
> 
> Two other things:
> 
> 1. Could you please clean up your commit history on this PR, by doing an interactive rebase (`git rebase -i HEAD~7`) and squashing all the commits in to a single commit with a clean commit message? See [here](https://docs.opentitan.org/doc/ug/github_notes/) for how you should structure your commit message, and [here](https://github.com/lowRISC/opentitan/pull/11028/commits/a43521cc36bec57ba08c5a0ca1b68edf0f5037ff) for an actual example.

One last thing, if you add the text `fixes #10852` in your commit message, the [corresponding issue will automatically be closed](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) when this PR merges

**Comment by MaayanKashani** [comment]
> Thanks for submitting this @maayanKash ! I just added a few minor cosmetic suggestions, otherwise overall LGTM.
> 
> We will probably want to add this same support to the bazel build as well; maybe @drewmacrae has some suggestions on how to accomplish this? While adding bazel support probably makes sense to do in a separate PR, @maayanKash would you mind filing an issue related to this, so we can track it?

Hi, @timothytrippel ,
Thanks for reviewing,
Will handle all cosmetics and uploads,
opened issue #11133 for Bazel support.

**Comment by MaayanKashani** [comment]
Hi, @timothytrippel ,
I'm new to Git.
Tried to rebase all changes, but due to pulling requests, I got confused.
All updated files are OK, how can I rebase them all now?

**Comment by timothytrippel** [comment]
> Hi, @timothytrippel , I'm new to Git. Tried to rebase all changes, but due to pulling requests, I got confused. All updated files are OK, how can I rebase them all now?

Hi @maayanKash, oh my, somehow your PR ended up with 37 commits! hehe not to worry, there are three ways we could fix this:

1. doing a somewhat complex series of `git rebase -i HEAD~<some # of commits>`, dropping / squashing, fixing rebase conflicts, and finally force pushing to this branch to update this PR, or

2. start a fresh branch off of master and manually copy over your changes to this new branch (this PR is small so there are only a few), create a single commit, and then create a new pull request.

or

3. I can fix it for you if you give me permission my selecting the checkbox shown below that says "Allow edits and access to secrets by maintainers".

Screen Shot 2022-03-01 at 10 16 16 PM

Since you are new to git, and I have already fixed it locally for you, I would suggest option 3, I just need permissions to push it.

Going forward, I recommend reading  [GitHub Notes](https://docs.opentitan.org/doc/ug/github_notes/) on the opentitan.org website, pretty thoroughly. It walks you through things like:
1. How to create a new development branch
2. How to write a good commit message
3. How to synchronize your forked repo with the upstream repo
4. How to to make changes to an outstanding PR, by amending prior commits and force-pushing to your PR branch
5. How to rebase changes on different branches
6. How to resolve rebase conflicts

Additionally, I also recommend this article: ["Git Tools - Rewriting History"](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History).

Let me know if you have any questions!

**Comment by timothytrippel** [comment]
Ok sounds good. No worries, once you open a new PR, I will close this one.

**Comment by timothytrippel** [comment]
Closing this as a new PR (#11188) has replaced this.

**Pull request closed without merging**